Be more patient in SSL tests

mpg opened the pull request:
## Description
We get too many spurious failures in the CI caused by not waiting long enough
for some things on the highly loaded CI machines. Just wait more; this doesn't
impact the quality of testing, just the time it takes to run the tests.

## Requires Backporting
Yes
[ ] Mbed TLS 2.1
[ ] Mbed TLS 1.3

Patater commented on tests/compat.sh:
```diff
@@ -888,7 +888,7 @@ if type lsof >/dev/null 2>/dev/null; then
 else
     wait_server_start() {
-        sleep 1
+        sleep 2
```
Is 2 seconds good enough for buildbot? Why not 3 seconds or 5 seconds?

mpg replied:
The (questionable) reasoning is: slow start issues, though frequent enough to be annoying, are not that systematic, so it could indicate `1` was already nearly enough, so I just doubled it.

Now I'm not opposed to making it large because, now that I think about it, impatient devs should have lsof, so this delay won't annoy them, and for the CI, slower with fewer false positives is probably better than less slow with more false positives, so why not 3 seconds indeed.

gilles-peskine-arm replied:
On developer machines, the `lsof` case will be used. And we're trying to get lsof installed on all CI. Would it be hard to install lsof on all buildbot instances?

Patater commented on tests/ssl-opt.sh:
```diff
@@ -656,14 +656,32 @@ fi
-    START_DELAY=1
-    DOG_DELAY=10
+    START_DELAY=2
+    DOG_DELAY=15
```
Why 15 and not 20?

mpg replied:
Because I was under the impression that the only spurious timeout we had so far where with valgrind, which would tend to indicate 10 was already enough, so I though +50% was enough here.

Now that you mention it, this has no impact on the normal running time of the script (by normal, I mean the client never hangs), so we could actually make the value much larger (like 1 min normal, 2 mins valgrind) just to be super-safe. (The only impact is that if the client is broken in such a way that it always hangs, then the running time of `ssl-opt.sh` will be around 300 * `$DOG_DELAY`, but this should probably be addressed at another level in the CI.)

Patater commented on tests/ssl-opt.sh:
```diff
@@ -656,14 +656,32 @@ fi
+    START_DELAY=2
+    DOG_DELAY=15
+    # allow impatient devs to override this one (really impacts running time)
+    : ${SRV_DELAY_BASE:=1}
```
Impatient devs are likely to have lsof. What's the impact when impatient devs have lsof?

mpg replied:
Unless I'm missing something, lsof only saves us from `$START_DELAY`, not from `$SRV_DELAY_BASE`: that one is intended to give the server some time to finish its thing after the client disconnected, including printing all the post-connection messages, and I don't think there is an easy way to check it has, so here we're just really waiting what we hope is a sufficient amount, and so far we have no better alternative in the script.

So actually without lsof you have a minimum of `$START_DELAY + $SRV_DELAY_BASE` spent just sleeping on each test; with lsof it's `epsilon + $SRV_DELAY_BASE`, so that's the case where `$SRV_DELAY_BASE` is the dominant one and you really want to make its value lower.

Patater replied:
OK. Sounds good to make this consistently configurable (valgrind or not). I can imagine a developer getting annoyed that his SRV_DELAY_BASE isn't being obeyed when he expects it to.

Patater commented on tests/ssl-opt.sh:
```diff
@@ -656,14 +656,32 @@ fi
-    DOG_DELAY=30
+    START_DELAY=6
+    DOG_DELAY=60
+    SRV_DELAY_BASE=2
```
Should impatient devs be able to override the SRV_DELAY_BASE when valgrind is being used?

mpg replied:
Impatient devs on Linux are likely to prefer ASan plus MemSan to valgrind, because even with the need to do two builds, it ends up being faster (and last time I checked, checking more) that valgrind's memcheck.

Impatient devs on OS X are out of luck because MemSan is still not available on non-Linux, and last time I tried (but this was two years ago), `ssl-opt.sh --memcheck` didn't work on OS X because apparently processes running under valgrind where unable to receive/catch signals, and we rely on the server cleaning up after receiving SIGTERM for our memory tests in `ssl-opt.sh` and `compat.sh`.

Now, of course it wouldn't hurt to make that overrideable too, if you think it can be useful (or better for consistency).

gilles-peskine-arm replied:
Impatient devs have been known to edit those timeouts down while they were developing. Ideally the scripts would be smarter and timeouts would only be needed in case a bug leads to an infinite loop. But I think this ideal world will wait for the day we rewrite those scripts in Python.

Patater commented on tests/ssl-opt.sh:
```diff
@@ -656,14 +656,32 @@ fi
+
+# some particular tests need more time:
+# - for the client, we multiply the usual delay by a factor
+# - for the severs, we add a number of secondes to the usual delay
```
s/secondes/seconds

mpg replied:
Haha, I can't hide I'm French, can I? I often find my fingers typing in the wrong language.

gilles-peskine-arm commented on tests/compat.sh:
```diff
@@ -888,7 +888,7 @@ if type lsof >/dev/null 2>/dev/null; then
     }
 else
     wait_server_start() {
```
I had a one-line patch to make diagnosis easier, it could be merged together with this PR: #1260

mpg commented:
@Patater @gilles-peskine-arm I've pushed fixes addressing your feedback, cherry-picked the warning from #1260 and optimised an easy common case. Please review again.

gilles-peskine-arm commented on tests/compat.sh:
```diff
@@ -887,8 +887,9 @@ if type lsof >/dev/null 2>/dev/null; then
         done
     }
 else
+    echo "Warning: lsof not available, wait_server_start = sleep 1"
```
Please change this to `… = sleep 2` or to `… = sleep` (displaying the timeout value isn't important, but if it's displayed, it should be correct).

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -308,6 +308,7 @@ if type lsof >/dev/null 2>/dev/null; then
         done
     }
 else
+    echo "Warning: lsof not available, wait_server_start = sleep 1"
```
Please change this to `… = sleep $START_DELAY`

mpg replied:
Haha, that's funny, I remember thinking that when looking at your PR, but then forgot about it when cherry-picking. Thanks for noticing, I'll fix that.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -340,6 +341,23 @@ check_server_hello_time() {
+    # the server still doesn't know the client has disconnected at the end).
+    # However the shortcut test is designed to only have false negatives and
+    # no false positives, and we fall back to sleeping on negatives.
+    if tail -n-1 $SRV_OUT | grep -F 'Waiting for a remote connection' >/dev/null; then :;
```
Technically this shortcut isn't actually correct: what if the server hasn't flushed its logs since it displayed the “Waiting” message?

mpg replied:
In principle, I agree with you. In practice, if the case you mentioned happened, then we would be seeing `server or client failed to reach handshake stage` errors (from around line 500) which I don't remember seeing. In general, I think the important point is that "shortcut + delay" is a strict improvement over the existing, ie all case that still fail after it also failed before.

However, I'll prepend `grep "Performing the SSL/TLS handshake" $SRV_OUT` to the shorcut condition for the sake of correcness.

gilles-peskine-arm replied:
I'm ok with the current code. It isn't perfect, but it's good enough for what it does, which is to shorten a timeout that is itself imperfect.

I'm not sure if grepping “Performing the SSL/TLS handshake” is really better: there are a few tests that perform multiple connections to the server.

There's no perfect solution anyway. In theory, waiting for the server to flush its logs could take an arbitrary time, which is indistinguishable from a server that's stuck in an infinite loop. What matters is that it works in practice on CI systems, and if we aren't seeing truncated logs any longer then we're all good.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -340,6 +341,23 @@ check_server_hello_time() {
+    # no false positives, and we fall back to sleeping on negatives.
+    if tail -n-1 $SRV_OUT | grep -F 'Waiting for a remote connection' >/dev/null; then :;
+    else
+        sleep $(( $SRV_DELAY_BASE + $SRV_DELAY_SECONDS ))
```
That's a very long timeout just for writing logs.

mpg replied:
I agree, and part of me was tempted to set `SRV_DELAY_BASE` to `0.1` instead (the other one already is 0 most of the time), but then, we keep seeing deadlines being missed on the CI machines... At this point I guess I just want to drastically reduce the false negative rate on the CI, and only after reaching that goal, work on reducing running time while keeping the false negative rate as close to 0 as possible. So IMO when we need to pick a somewhat arbitrary timeout like here (which hopefully we can get rid of in the future), I'd rather be too conservative than not enough for now.

So unless you strongly disagree, I'm inclined to keep this one to 1 full second.

gilles-peskine-arm replied:
I'm not sure if `sleep` supports fractional arguments on all the platforms we want to support.

mpg replied:
I might be missing something, but feel like I'm getting contradictory signals here: first you were noticing that 1 second is long, now you seem to be arguing against fractional arguments, and unless you have something else than `sleep` in mind, I don't see how these two remarks can be reconciled into something actionable. Can you clarify if you'd like be to change something here?

gilles-peskine-arm replied:
Oh, sorry, I'm the one who was missing something. I was reading the wrong values and thought that the total was much higher than 1.

But on re-reading this, I think I'm missing something else. Why do this check _before_ sending a signal to the server? If the server is operating normally, then `kill $SRV_PID; wait $SRV_PID` should result in {{ssl_server2}} terminating normally (after running its SIGTERM handler). So shouldn't any timeout be on the `wait`, in case the server is stuck in an infinite loop?

mpg replied:
I don't think we're trying to address the case where the server is stuck in an infinite loop here: so far we assume it'll always terminate nicely after receiving SIGTERM and I don't think we've noticed the opposite happening so far (fingers crossed).

What this delay was trying to solve is that I suspected that sometimes the server was killed before it had time to write the expected "Last error was ..." message. But now that you comment made me reconsider that, I realised that we shouldn't need any delay for that: we're in control of what the server does when receiving SIGTERM, so it's up to us to ensure that if does print everything it need to print on its way out.

So I think this arbitrary sleep was entirely the wrong way of addressing the problem, and I'm now inclined to withdraw it - see my new take on IOTSSL-2006 on our internal tracker, and thanks for forcing me to clarify my thinking here..

mpg commented:
@Patater @gilles-peskine-arm This PR tried to address two issues (slow start and incomplete server log), but I later realized only the first one (slow start) should be addressed in the test scripts, the other on is in `ssl_server2.c`. So I'm closing this PR and opening https://github.com/ARMmbed/mbedtls/pull/1286 for slow start (please review), and will open a separate one shortly for incomplete server logs.
